WT-15693 Verify delta chain in PALite

ablekhman opened the pull request:
1. Verify delta chain integrity in the following calls:
- `plh_put`
- `plh_get`
- `plh_discard`
2. Add documentation for selected queries (work in progress, more to follow)

ablekhman committed:
WT-15693 Verify delta chain in PALite
Verify delta chain integrity in the following calls:
- `plh_put`
- `plh_get`
- `plh_discard`

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -576,6 +580,34 @@ db_busy_handler(void *ptr, int count)
 // Storage layer
 //
 
+struct PageInfo {
```
This is not a new code. Just moved here from below so `std::formatter<PageInfo>` can find the type.

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -585,13 +617,14 @@ class Storage {
         PUT_CHECKPOINT,
         GET_CHECKPOINT,
-        DELETE_CHECKPOINT,
+        DELETE_CHECKPOINTS,
```
Changed few queries to plural to reflect what they actually do.

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -681,29 +713,37 @@ class Storage {
                         lsn DESC,
                         timestamp DESC
                     LIMIT 1;)";
-                /* The sqlite bind call takes a signed 64 bit integer.
```
Added `WT_PAGE_LOG_LSN_MAX` as a query argument, so no explanation needed.

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -717,20 +757,113 @@ class Storage {
                         );)";
-                a[GET_PAGE] = R"(
+
+                /* !!! Retrieve information about entire delta chain stopping at
```
I added such detailed explanation of the query because the query is not trivial. If it causes too much confusion and performance degradation with a simpler query is within acceptable range, then we may reconsider it.

The simpler query would be something like this (some fields omitted for clarity):

```sql
WITH terminating_page AS (
  SELECT MAX(lsn) AS lsn
  FROM pages
  WHERE lsn <= :lsn AND delta = 0
)
SELECT
  p.lsn,
  p.page_id,
  p.flags,
  p.backlink_lsn,
  p.base_lsn
FROM
  pages p,
  terminating_page tp
WHERE
  p.lsn <= :lsn AND
  (tp.lsn IS NULL OR p.lsn >= tp.lsn)
ORDER BY
  p.lsn DESC;
```

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1123,51 +1296,77 @@ class Storage {
 
     int
-    get_page(Connection &conn, uint64_t table_id, uint64_t page_id, WT_PAGE_LOG_GET_ARGS *args,
+    get_page_infos(Connection &conn, uint64_t table_id, uint64_t page_id, uint64_t lsn,
```
Only retrieve pages information for verification. Page data omitted.

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1184,60 +1383,142 @@ class Storage {
+        std::for_each(pages.rbegin(), pages.rend(), verify_page);
+
+        // Verify discarded page separately because it can backlink to any page in the chain.
+        if (discarded.lsn != 0) {
```
I'm not 100% sure about why discarded pages may backlink to any page in the chain. I ported it from PALM as it is. If this is not needed any longer, I'm happy to remove this check.

pmacko86 replied:
If you are curious, feel free to ask Chenhao. I don't remember the reason, but it sounded like it would be quite hard to make it always point to the latest page - and given that the storage infrastructure underneath us doesn't care that it's pointing to the last page, it wasn't worth fixing.

ablekhman commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1445,7 +1731,7 @@ class Palite : public WT_PAGE_LOG {
 
     // Reference counting for the page log service
-    int ref_count;
+    std::atomic_int ref_count;
```
Related to similar fix in PALM: WT-15678. Apparently, the counter can be contended.

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -859,14 +993,51 @@ class Storage {
+          The primary key for the table (table_id, page_id, lsn) does not help
+          because lsn is always increasing, making each record unique.
+
+          Including backlink_lsn in the primary key would not work because
```
Did you mean `base_lsn` in this paragraph? `backlink_lsn` would be unique, and it would *not* be 0 for the full page image (except for the first full page image of the page).

ablekhman replied:
I changed the comment to explain why `backlink_lsn` does not fit for primary index. In short, `backlink_lsn` cannot be in the same index with `lsn` because together they always form a unique pair. I'm interested to catch the case when `backlink_lsn` is not unique within delta chain.

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -859,14 +993,51 @@ class Storage {
              PRIMARY KEY (table_id, page_id, lsn)
             );)",
+
+          /* !!! To gracefully handle write failures of delta pages we will keep
```
Why do we need this functionality?

With the abandon checkpoint functionality implemented, this should never happen.

ablekhman replied:
I talked to Chenhao about it. He says it's still in use.

pmacko86 replied:
Can you please elaborate on this? Why do we need this? This cannot be correct.

There should not be any case in which we have written a page delta and then need to write it again. The only exception is if we crash, in which case we handle it by abandoning the checkpoint - there shouldn't be any need to handle this elsewhere. In disagg, there is no other way to recover from failed writes other than to panic, so there should be no provision for graceful recovery.

PALM doesn't support this either. In fact, there is even a check in plh_put that verifies whether the backlink_lsn is set to the previous page version (for deltas), and it would assert if you tried to re-write the page with the same backlink_lsn. (In fact, we should port that check to PALite.)

quchenhao replied:
This generally happens in eviction. We have a base page already written and we now try to evict it. We finished writing a delta to disk but then we fail the reconciliation. It is a delta so we cannot explicitly free the write with the same page id. After a while, we retry the eviction and this time we writes a delta based on the same base page. Thus this write will have the same backlink lsn to the previous write.

At this stage, there may be no checkpoint running so discarding the unfinished checkpoint doesn't help in this case. It is also possible that these two writes will be included in the same checkpoint that is not discarded.

pmacko86 replied:
That's quite unfortunate. Under which circumstances would the reconciliation fail after writing the delta? I thought this was after the point in the code that says that any failure after that point would result in panic. And I also thought that this could only happen with tombstones - not realizing that this could also happen with other deltas.

Do we have this described somewhere in a ticket? Even if this is something that the storage infrastructure could live with, it still feels like something that would be worth cleaning up in the future if we can. If we could clean this up, it would get us tighter and clearer semantics with fewer special cases, which means that it would eliminate some complexity elsewhere and improve our ability to test it more thoroughly. If we don't have a ticket for it yet, @quchenhao, could you please create it? Even if it would be labeled as P4 "Tech Debt," I think it would be still worth it.

Either way, we need an explanation in the comments about why we have this behavior, as without it, it is very much unexpected. @ablekhman, could you please incorporate a summary of Chenhao's explanation into the comment in the code? After that, we would be good to go here.

quchenhao replied:
I can create a ticket for that. The issue mainly happens when the reconciliation doesn't make any progress.

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -859,14 +993,51 @@ class Storage {
+
+          /* !!! To gracefully handle write failures of delta pages we will keep
+          track of the backlink_lsn within each delta chain. Tracking is done via
+          a partial unique index on (table_id, page_id, backlink_lsn) for delta
```
We don't guarantee this triplet to be unique, even in the case when all writes are successful. There is some weird condition that I don't really understand, under which a tombstone write's `backlink_lsn` may not refer to the page immediately before it. (Please refer to Yury's PALM code that verifies tombstone backlink LSNs.)

ablekhman replied:
Good point. Discarded pages should be excluded from this index, even though they are technically deltas.

pmacko86 replied:
There is actually another case in which this triplet wouldn't be unique: This is because the backlink_lsn of a full page needs to point to the previous full page.

For example, consider this sequence:
* Full page, LSN = 10, backlink = 0, base = 0
* Delta page, LSN = 11, backlink = 10, base = 10
* Delta page, LSN = 12, backlink = 11, base = 10
* Full page, LSN = 13, backlink = 10, base = 0

Note that backlink_lsn = 10 appears twice.

ablekhman replied:
This case is handled by excluding everything but genuine delta pages from the index. In the example above, the index will contain (in bold, omitting table_id for clarity):

Delta page, LSN = 11, **backlink = 10, page_id=42**, base = 10
Delta page, LSN = 12, **backlink = 11, page_id=42**, base = 10

pmacko86 replied:
Ok, that makes sense - if it's just deltas, this should work. Please just make sure it's explained in the comment, if it's not there already.

@quchenhao, just to make sure, the problem with failed reconciliation could happen only with deltas, but not full page images? Or is there a possibility that we'd write a delta, fail, and then replace it with a full page image?

quchenhao replied:
```
    /*
     * If eviction didn't use any updates and didn't split or delete the page, it didn't make
     * progress. Give up rather than silently succeeding in doing no work: this way threads know to
     * back off forced eviction rather than spinning.
     *
     * Do not return an error if we are syncing the file with eviction disabled or as part of a
     * checkpoint.
     */
    if (ret == 0 && !(btree->evict_disabled > 0 || !F_ISSET(btree->dhandle, WT_DHANDLE_OPEN)) &&
      F_ISSET(r, WT_REC_EVICT) && !WT_PAGE_IS_INTERNAL(page) && r->multi_next == 1 &&
      F_ISSET(r, WT_REC_CALL_URGENT) && !r->update_used && r->cache_write_restore_invisible &&
      !r->cache_upd_chain_all_aborted) {
        /*
         * If leaf delta is enabled, we should have built an empty delta if this page has been
         * reconciled before as we don't make any progress.
         */
        WT_ASSERT(session,
          !WT_DELTA_ENABLED_FOR_PAGE(session, page->type) || F_ISSET(r, WT_REC_EMPTY_DELTA) ||
            page->disagg_info->block_meta.page_id == WT_BLOCK_INVALID_PAGE_ID);
        /*
         * If eviction didn't make any progress, let application threads know they should refresh
         * the transaction's snapshot (and try to evict the latest content).
         */
        if (F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT))
            F_SET(session->txn, WT_TXN_REFRESH_SNAPSHOT);

        WT_STAT_CONN_DSRC_INCR(session, cache_eviction_blocked_no_progress);
        ret = __wt_set_return(session, EBUSY);
    }
```

I made some changes to ensure we only write empty deltas if we don't make any progress. So I think this issue mostly has gone away in production. But we have introduced a fail point that simulates the failure behaviour.

```
    /*
     * Fail 1% of the time after we have built the disk image but before we wrap up reconciliation.
     */
    if (F_ISSET(r, WT_REC_EVICT) && !F_ISSET(r, WT_REC_EVICT_CALL_CLOSING) &&
      __wt_failpoint(session, WT_TIMING_STRESS_FAILPOINT_REC_BEFORE_WRAPUP, 100))
        ret = __wt_set_return(session, EBUSY);
```
So I expect this issue will still happen in testing.

> we'd write a delta, fail, and then replace it with a full page image

Yes, that's possible.

pmacko86 replied:
@ablekhman, does PALite handle the case in which a failed delta write would be replaced by a full page image? It seems to me that it would currently only do it when it is being replaced by a delta.

If this still needs to get implemented, feel free to do it in a separate PR—just please create a ticket for it, so that we won't forget.

ablekhman replied:
Chenhao explained to me that a new full page will have different `page_id`. So there is no conflict with the existing delta.
However, this failed delta write will stay in the DB because at PALite level it is indistinguishable from any other delta. Only the app knows that it "failed". If this is a problem, then we may need to provide an API to clean bad deltas from DB.

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1184,60 +1383,142 @@ class Storage {
+        // Always verify the delta chain when retrieving pages, even if config.verify=false.
+        verify_chain(pages);
+
+        // Fill args from the full page.
```
I thought we need to do this based on the last version of the page, not the full page image? I guess we should double-check what PALM does.

ablekhman replied:
Good catch!

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1184,60 +1383,142 @@ class Storage {
+        PageInfo base{};
+        PageInfo discarded{};
+
+        auto verify_base = [&](const PageInfo &page) {
```
Please rename these to "full" pages instead of "base" pages, here and also in the log messages below.

ablekhman replied:
Will do.

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1184,60 +1383,142 @@ class Storage {
+                LOG_AND_THROW("Multiple base pages in chain: {}, {}", base, page);
+            }
+
+            if (page.backlink_lsn != 0) {
```
This isn't right. The `backlink_lsn` of a full page points to the previous full page image.

ablekhman replied:
I'll fix it.

pmacko86 reviewed (commented):
Thanks for the PR! It's almost there, but there are a few issues that I mentioned in the comments.

ablekhman committed:
WT-15693 Verify delta chain in PALite
Verify delta chain integrity in the following calls:
- `plh_put`
- `plh_get`
- `plh_discard`

pmacko86 commented on ext/page_log/palite/palite.cpp:
```diff
@@ -1184,60 +1408,138 @@ class Storage {
+
+    Returns 0 on success or throws on error. */
+    int
+    verify_page_chain(Connection &conn, uint64_t table_id, uint64_t page_id, uint64_t lsn)
```
To fully match the checks in PALM, please add the check that verifies that the full page image's backlink_lsn is set correctly to the previous full page image. Feel free to do this in a separate ticket to avoid the scope creep.

(PALM does this only during put, but it wouldn't hurt to be able to validate this elsewhere—either way.)

ablekhman replied:
Created a ticket to track this task: WT-15794

ablekhman committed:
WT-15693 Verify delta chain in PALite
Verify delta chain integrity in the following calls:
- `plh_put`
- `plh_get`
- `plh_discard`

pmacko86 reviewed (approved):
LGTM
